File: crm/app/internal/adapters/inbound/http/conversation_handler.go
// Package http — Conversation/inbox + Messenger ingest HTTP handlers.
// Auth: DEFERRED — LAN-trust only (per plan). FB webhook signature verification also deferred.
package http

import (
	"context"
	"encoding/json"
	"io"
	"log"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/vantt/data-integration/crm/app/internal/adapters/inbound/messenger"
	"github.com/vantt/data-integration/crm/app/internal/domain"
)

// InboxQuerier is the read interface for conversations.
type InboxQuerier interface {
	ListInbox(ctx context.Context, assigneeUserID, status string) ([]domain.Conversation, error)
	GetMessages(ctx context.Context, conversationID string) ([]domain.Message, error)
}

// InboxWriter covers conversation assignment and status changes.
type InboxWriter interface {
	AssignConversation(ctx context.Context, conversationID, assigneeUserID string) error
	SetStatus(ctx context.Context, conversationID, newStatus string) error
}

// MessageIngester accepts parsed chat messages.
type MessageIngester interface {
	IngestMessage(ctx context.Context, msg domain.ParsedMessage) error
}

// ConversationHandler wires inbox and ingest endpoints.
type ConversationHandler struct {
	querier  InboxQuerier
	writer   InboxWriter
	ingester MessageIngester
}

// NewConversationHandler constructs the handler.
func NewConversationHandler(q InboxQuerier, w InboxWriter, ing MessageIngester) *ConversationHandler {
	return &ConversationHandler{querier: q, writer: w, ingester: ing}
}

// RegisterRoutes mounts conversation routes on the given chi router.
func (h *ConversationHandler) RegisterRoutes(r chi.Router) {
	r.Get("/api/conversations", h.listInbox)
	r.Patch("/api/conversations/{id}", h.updateConversation)
	r.Get("/api/conversations/{id}/messages", h.listMessages)
	// Ingest endpoint: accepts a FB Messenger webhook payload for testing/dev.
	// Auth/signature verification: TODO when live FB token is wired.
	r.Post("/api/messenger/ingest", h.messengerIngest)
}

// GET /api/conversations?assignee=<user_id>&status=<status>
func (h *ConversationHandler) listInbox(w http.ResponseWriter, r *http.Request) {
	assignee := r.URL.Query().Get("assignee")
	status := r.URL.Query().Get("status")
	convs, err := h.querier.ListInbox(r.Context(), assignee, status)
	if err != nil {
		log.Printf("conversation handler: list inbox: %v", err)
		writeError(w, http.StatusInternalServerError, "failed to list conversations")
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"conversations": convs})
}

// PATCH /api/conversations/{id}   body: {"assignee_user_id":"..."} or {"status":"closed"}
func (h *ConversationHandler) updateConversation(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "id")

	var body struct {
		AssigneeUserID *string `json:"assignee_user_id"`
		Status         *string `json:"status"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if body.AssigneeUserID != nil {
		if err := h.writer.AssignConversation(r.Context(), convID, *body.AssigneeUserID); err != nil {
			if domain.IsValidationError(err) {
				writeError(w, http.StatusBadRequest, err.Error())
			} else {
				log.Printf("conversation handler: assign %s: %v", convID, err)
				writeError(w, http.StatusInternalServerError, "operation failed")
			}
			return
		}
	}
	if body.Status != nil {
		if err := h.writer.SetStatus(r.Context(), convID, *body.Status); err != nil {
			if domain.IsValidationError(err) {
				writeError(w, http.StatusBadRequest, err.Error())
			} else {
				log.Printf("conversation handler: set status %s: %v", convID, err)
				writeError(w, http.StatusInternalServerError, "operation failed")
			}
			return
		}
	}
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok", "conversation_id": convID})
}

// GET /api/conversations/{id}/messages
func (h *ConversationHandler) listMessages(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "id")
	msgs, err := h.querier.GetMessages(r.Context(), convID)
	if err != nil {
		log.Printf("conversation handler: list messages %s: %v", convID, err)
		writeError(w, http.StatusInternalServerError, "failed to list messages")
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"conversation_id": convID, "messages": msgs})
}

// POST /api/messenger/ingest
// Accepts a raw FB Messenger webhook payload JSON.
// Parses it and ingests each message event.
// Auth/FB signature verification: TODO — deferred until live token is available.
func (h *ConversationHandler) messengerIngest(w http.ResponseWriter, r *http.Request) {
	raw, err := io.ReadAll(io.LimitReader(r.Body, 1<<20)) // 1 MiB cap
	if err != nil {
		writeError(w, http.StatusBadRequest, "failed to read body")
		return
	}

	msgs, err := messenger.ParseWebhookPayload(raw)
	if err != nil {
		log.Printf("conversation handler: parse messenger payload: %v", err)
		writeError(w, http.StatusBadRequest, "invalid messenger payload")
		return
	}

	ingested, skipped := 0, 0
	for _, msg := range msgs {
		if err := h.ingester.IngestMessage(r.Context(), msg); err != nil {
			log.Printf("conversation handler: ingest message %s: %v", msg.ExternalMessageID, err)
			skipped++
			continue
		}
		ingested++
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"parsed":   len(msgs),
		"ingested": ingested,
		"skipped":  skipped,
	})
}

File: crm/src/adapters/inbound/http/conversation_handler.py
"""FastAPI router: Conversation inbox + Messenger ingest endpoints.

Auth:
  PATCH /api/conversations/{id}  — requires X-CRM-Token (CRM_API_TOKEN env var).
  GET endpoints                   — unauthenticated (read-only).
  POST /api/conversations/messenger/ingest — NO CRM token (external FB webhook;
      Facebook signs payloads with X-Hub-Signature-256, not our internal token).
"""
from __future__ import annotations

import logging
from dataclasses import asdict
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel

from adapters.inbound.http.auth_dependency import require_api_token
from application.conversation_service import ConversationService
from application.messenger_parser import parse_messenger_webhook

log = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Request/response schemas
# ---------------------------------------------------------------------------

class UpdateConversationBody(BaseModel):
    assignee_user_id: Optional[str] = None
    status: Optional[str] = None


# ---------------------------------------------------------------------------
# Handler factory
# ---------------------------------------------------------------------------

def make_conversation_router(conv_svc: ConversationService) -> APIRouter:
    """Return a configured APIRouter with all conversation endpoints wired."""
    router = APIRouter(prefix="/api")

    # GET /api/conversations?assignee=<user_id>&status=<status>
    @router.get("/conversations")
    def list_conversations(
        assignee: str = "",
        status: str = "",
    ):
        convs = conv_svc.list_inbox(assignee_user_id=assignee, status=status)
        return {"conversations": [asdict(c) for c in convs]}

    # PATCH /api/conversations/{id}
    @router.patch("/conversations/{conversation_id}", dependencies=[Depends(require_api_token)])
    def update_conversation(
        conversation_id: str,
        body: UpdateConversationBody,
    ):
        if body.assignee_user_id is not None:
            try:
                conv_svc.assign_conversation(conversation_id, body.assignee_user_id)
            except ValueError as exc:
                raise HTTPException(status_code=400, detail=str(exc))
            except Exception as exc:
                log.error("assign conversation %s: %s", conversation_id, exc)
                raise HTTPException(status_code=500, detail="operation failed")

        if body.status is not None:
            try:
                conv_svc.set_status(conversation_id, body.status)
            except ValueError as exc:
                raise HTTPException(status_code=400, detail=str(exc))
            except Exception as exc:
                log.error("set status %s: %s", conversation_id, exc)
                raise HTTPException(status_code=500, detail="operation failed")

        return {"status": "ok", "conversation_id": conversation_id}

    # GET /api/conversations/{id}/messages
    @router.get("/conversations/{conversation_id}/messages")
    def list_messages(conversation_id: str):
        msgs = conv_svc.list_messages(conversation_id)
        return {"conversation_id": conversation_id, "messages": [asdict(m) for m in msgs]}

    # POST /api/conversations/messenger/ingest
    # Accepts a raw FB Messenger webhook payload JSON.
    # Auth/FB signature verification: TODO — deferred until live token is available.
    @router.post("/conversations/messenger/ingest")
    async def messenger_ingest(request: Request):
        body = await request.body()
        if len(body) > 1 << 20:  # 1 MiB cap
            raise HTTPException(status_code=413, detail="payload too large")

        try:
            payload = __import__("json").loads(body)
        except Exception:
            raise HTTPException(status_code=400, detail="invalid JSON body")

        try:
            parsed = parse_messenger_webhook(payload)
        except Exception as exc:
            log.error("parse messenger payload: %s", exc)
            raise HTTPException(status_code=400, detail="invalid messenger payload")

        ingested, skipped = 0, 0
        for msg in parsed:
            try:
                conv_svc.ingest_message(msg)
                ingested += 1
            except Exception as exc:
                log.error("ingest message %s: %s", msg.external_message_id, exc)
                skipped += 1

        return {"parsed": len(parsed), "ingested": ingested, "skipped": skipped}

    return router
